File: app/src/main/java/com/streamarc/tv/ui/TransferDialogs.kt
package com.streamarc.tv.ui

import android.Manifest
import android.content.Intent
import android.net.Uri
import android.provider.Settings
import android.os.Build
import android.view.LayoutInflater
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.streamarc.tv.R
import com.streamarc.tv.data.Prefs
import com.streamarc.tv.databinding.DialogRecordBinding
import com.streamarc.tv.transfer.Folders
import com.streamarc.tv.transfer.TransferJob
import com.streamarc.tv.transfer.TransferService
import com.streamarc.tv.transfer.TransferType
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.UUID

/** One thing to download: title for the list, URL and file extension. */
data class DownloadItem(val title: String, val subtitle: String, val url: String, val ext: String, val fileBase: String? = null)

object TransferDialogs {

    /**
     * Queues every item into the saved download folder. The picker only appears the first time
     * (or if the saved folder is no longer reachable); the folder is changed in Settings.
     * Falls back to the app folder on devices without a picker.
     */
    fun download(activity: AppCompatActivity, picker: FolderPicker, items: List<DownloadItem>) {
        if (items.isEmpty()) return
        val prefs = Prefs(activity)
        val current = prefs.downloadFolder?.takeIf { Folders.usable(activity, it) }
        if (current == null) {
            pickFolder(activity, picker, TransferType.DOWNLOAD) { folder -> enqueueDownloads(activity, items, folder) }
            return
        }
        enqueueDownloads(activity, items, current)
    }

    /** Opens the picker, saves the choice for this type, and continues with it (null = app folder). */
    fun pickFolder(activity: AppCompatActivity, picker: FolderPicker, type: TransferType, then: (String?) -> Unit) {
        val prefs = Prefs(activity)
        val opened = picker.launch { uri ->
            if (uri != null) {
                val s = uri.toString()
                if (type == TransferType.RECORDING) prefs.recordingFolder = s else prefs.downloadFolder = s
                then(s)
            } else if ((if (type == TransferType.RECORDING) prefs.recordingFolder else prefs.downloadFolder) == null) {
                // Cancelled with nothing saved yet: use the app folder so the action still works.
                then(null)
            }
        }
        if (!opened) {
            Toast.makeText(activity, R.string.no_folder_picker, Toast.LENGTH_LONG).show()
            then(null)
        }
    }

    private fun enqueueDownloads(activity: AppCompatActivity, items: List<DownloadItem>, folder: String?) {
        ensureNotifications(activity)
        for (it in items) {
            val job = TransferJob(
                id = UUID.randomUUID().toString(),
                type = TransferType.DOWNLOAD,
                title = it.title,
                subtitle = it.subtitle,
                url = it.url,
                fileName = Folders.safeName(it.fileBase ?: it.title) + "." + it.ext.ifBlank { "mp4" },
                folder = folder,
                startAt = 0, endAt = 0,
                createdAt = System.currentTimeMillis()
            )
            TransferService.enqueue(activity, job)
        }
        val msg = if (items.size == 1) activity.getString(R.string.download_started_fmt, items[0].title)
        else activity.getString(R.string.downloads_started_fmt, items.size)
        // Queued: offer the Downloads screen, or stay put.
        AlertDialog.Builder(activity)
            .setTitle(R.string.download_queued_title)
            .setMessage(msg)
            .setPositiveButton(R.string.go_to_downloads) { _, _ ->
                activity.startActivity(TransfersActivity.intent(activity, TransferType.DOWNLOAD))
            }
            .setNegativeButton(R.string.stay_here, null)
            .show()
    }

    /** Start and end clock times (to the minute) on hour / minute / AM-PM wheels, then folder and schedule. */
    fun record(activity: AppCompatActivity, picker: FolderPicker, channelName: String, url: String) {
        val vb = DialogRecordBinding.inflate(LayoutInflater.from(activity))
        val now = com.streamarc.tv.data.Format.calendar()
        val startCal = (now.clone() as java.util.Calendar).apply { add(java.util.Calendar.MINUTE, 1) }
        val endCal = (startCal.clone() as java.util.Calendar).apply { add(java.util.Calendar.HOUR_OF_DAY, 1) }

        val start = TimeWheels(vb.startHour, vb.startMinute)
        val end = TimeWheels(vb.endHour, vb.endMinute)
        start.set(startCal.get(java.util.Calendar.HOUR_OF_DAY), startCal.get(java.util.Calendar.MINUTE))
        end.set(endCal.get(java.util.Calendar.HOUR_OF_DAY), endCal.get(java.util.Calendar.MINUTE))

        fun times(): Pair<Long, Long> = resolve(start.hour24(), start.minute(), end.hour24(), end.minute())
        fun summarize() {
            val (st, en) = times()
            val f = com.streamarc.tv.data.Format
            val sameDay = f.dayMs(st) == f.dayMs(en)
            vb.txtSummary.text = if (sameDay) "${f.dayMs(st)}   ${f.timeMs(st)} – ${f.timeMs(en)}"
            else "${f.dayMs(st)} ${f.timeMs(st)} – ${f.dayMs(en)} ${f.timeMs(en)}"
        }
        start.onChange = { summarize() }
        end.onChange = { summarize() }
        summarize()

        val prefs = Prefs(activity)
        vb.txtFolder.text = activity.getString(
            R.string.folder_current_fmt,
            Folders.describe(activity, prefs.recordingFolder?.takeIf { Folders.usable(activity, it) }, TransferType.RECORDING)
        )
        val dialog = AlertDialog.Builder(activity)
            .setTitle(activity.getString(R.string.record_fmt, channelName))
            .setView(vb.root)
            .setPositiveButton(R.string.record) { _, _ ->
                val (st, en) = times()
                val current = prefs.recordingFolder?.takeIf { Folders.usable(activity, it) }
                if (current == null && prefs.recordingFolder == null) {
                    pickFolder(activity, picker, TransferType.RECORDING) { f -> schedule(activity, channelName, url, st, en, f) }
                } else schedule(activity, channelName, url, st, en, current)
            }
            .setNegativeButton(android.R.string.cancel, null)
            .create()
        dialog.show()
        vb.startHour.requestFocus()
    }

    /**
     * One hour wheel that runs through the whole day (12 AM … 11 AM, 12 PM … 11 PM and
     * wraps to the next day) plus a minute wheel. Works by touch and with a remote.
     */
    private class TimeWheels(val hour: android.widget.NumberPicker, val minute: android.widget.NumberPicker) {
        var onChange: (() -> Unit)? = null

        init {
            hour.minValue = 0; hour.maxValue = 23; hour.wrapSelectorWheel = true
            hour.displayedValues = Array(24) { h ->
                val h12 = if (h % 12 == 0) 12 else h % 12
                "$h12 ${if (h < 12) "AM" else "PM"}"
            }
            minute.minValue = 0; minute.maxValue = 59; minute.wrapSelectorWheel = true
            minute.setFormatter { v -> String.format(Locale.US, "%02d", v) }
            for (p in listOf(hour, minute)) {
                p.descendantFocusability = android.view.ViewGroup.FOCUS_BLOCK_DESCENDANTS   // wheels only, no keyboard
                p.setOnValueChangedListener { _, _, _ -> onChange?.invoke() }
            }
        }

        fun set(hour24: Int, min: Int) { hour.value = hour24; minute.value = min }
        fun hour24(): Int = hour.value
        fun minute(): Int = minute.value
    }

    /**
     * Turns start/end clock times into epoch millis: a start earlier than now means
     * tomorrow (unless it is within the last minute, which means "now"), and an end
     * at or before the start rolls over to the next day.
     */
    private fun resolve(sh: Int, sm: Int, eh: Int, em: Int): Pair<Long, Long> {
        val now = System.currentTimeMillis()
        val start = com.streamarc.tv.data.Format.calendar().apply {
            set(java.util.Calendar.HOUR_OF_DAY, sh); set(java.util.Calendar.MINUTE, sm); set(java.util.Calendar.SECOND, 0); set(java.util.Calendar.MILLISECOND, 0)
        }
        var st = start.timeInMillis
        if (st < now - 90_000) st += 24 * 3600 * 1000L
        if (st < now) st = now
        val end = com.streamarc.tv.data.Format.calendar().apply {
            timeInMillis = st
            set(java.util.Calendar.HOUR_OF_DAY, eh); set(java.util.Calendar.MINUTE, em); set(java.util.Calendar.SECOND, 0); set(java.util.Calendar.MILLISECOND, 0)
        }
        var en = end.timeInMillis
        if (en <= st) en += 24 * 3600 * 1000L
        return st to en
    }

    private fun schedule(activity: AppCompatActivity, channel: String, url: String, startAt: Long, endAt: Long, folder: String?) {
        ensureNotifications(activity)
        val f = com.streamarc.tv.data.Format
        val stamp = f.fmt("yyyy-MM-dd HH.mm").format(Date(startAt))
        val job = TransferJob(
            id = UUID.randomUUID().toString(),
            type = TransferType.RECORDING,
            title = channel,
            subtitle = "${f.dayMs(startAt)} · ${f.timeMs(startAt)} – ${f.timeMs(endAt)}",
            url = url,
            fileName = Folders.safeName("$channel $stamp") + ".ts",
            folder = folder,
            startAt = startAt,
            endAt = endAt,
            createdAt = System.currentTimeMillis()
        )
        TransferService.enqueue(activity, job)
        val startsNow = startAt <= System.currentTimeMillis() + 60_000
        val msg = if (startsNow) activity.getString(R.string.recording_started_fmt, channel)
        else activity.getString(R.string.recording_scheduled_fmt, channel, f.timeMs(startAt))
        Toast.makeText(activity, msg, Toast.LENGTH_LONG).show()
        if (!startsNow) checkRecordingReadiness(activity)
    }

    /**
     * A scheduled recording needs Android to wake the app on time and to leave it running.
     * If exact alarms are blocked or battery optimisation is on, say so now and open the
     * setting, rather than finding an empty file in the morning.
     */
    fun checkRecordingReadiness(activity: AppCompatActivity) {
        val problems = ArrayList<Pair<String, Intent>>()
        val pkg = activity.packageName
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val am = activity.getSystemService(android.content.Context.ALARM_SERVICE) as android.app.AlarmManager
            if (!am.canScheduleExactAlarms()) {
                problems.add(activity.getString(R.string.readiness_exact_alarms) to
                    Intent(Settings.ACTION_REQUEST_SCHEDULE_EXACT_ALARM, Uri.parse("package:$pkg")))
            }
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val pm = activity.getSystemService(android.content.Context.POWER_SERVICE) as android.os.PowerManager
            if (!pm.isIgnoringBatteryOptimizations(pkg)) {
                val direct = Intent(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS, Uri.parse("package:$pkg"))
                val list = Intent(Settings.ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS)
                val intent = when {
                    direct.resolveActivity(activity.packageManager) != null -> direct
                    list.resolveActivity(activity.packageManager) != null -> list
                    else -> null
                }
                if (intent != null) problems.add(activity.getString(R.string.readiness_battery) to intent)
            }
        }
        if (problems.isEmpty()) return
        var i = 0
        fun showNext() {
            if (i >= problems.size) return
            val (text, intent) = problems[i++]
            AlertDialog.Builder(activity)
                .setTitle(R.string.readiness_title)
                .setMessage(text)
                .setPositiveButton(R.string.open_settings) { _, _ ->
                    try { activity.startActivity(intent) } catch (_: Exception) {
                        Toast.makeText(activity, R.string.readiness_no_screen, Toast.LENGTH_LONG).show()
                    }
                    showNext()
                }
                .setNegativeButton(R.string.not_now) { _, _ -> showNext() }
                .show()
        }
        showNext()
    }

    fun offsetLabel(minutes: Int): String {
        val h = minutes / 60; val m = minutes % 60
        return when {
            h == 0 -> "$m min"
            m == 0 -> "$h h"
            else -> "$h h $m min"
        }
    }

    private fun ensureNotifications(activity: AppCompatActivity) {
        if (Build.VERSION.SDK_INT >= 33 &&
            ContextCompat.checkSelfPermission(activity, Manifest.permission.POST_NOTIFICATIONS) != android.content.pm.PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(activity, arrayOf(Manifest.permission.POST_NOTIFICATIONS), 71)
        }
    }
}

File: windows/StreamArcTV/UI/TransferDialogs.cs
using System.Globalization;
using System.Windows;
using System.Windows.Controls;
using StreamArcTV.Data;
using StreamArcTV.Transfer;

namespace StreamArcTV.UI;

/// One thing to download: title for the list, URL and file extension.
public record DownloadItem(string Title, string Subtitle, string Url, string Ext);

public static class TransferDialogs
{
    /// <summary>
    /// Queues every item into the saved download folder. The picker only appears the first time
    /// (or if the saved folder is no longer reachable); the folder is changed in Settings.
    /// </summary>
    public static void Download(IList<DownloadItem> items)
    {
        if (items.Count == 0) return;
        var prefs = Prefs.Instance;
        var current = prefs.DownloadFolder != null && Folders.Usable(prefs.DownloadFolder) ? prefs.DownloadFolder : null;
        if (current == null)
        {
            PickFolder(TransferType.DOWNLOAD, folder => EnqueueDownloads(items, folder));
            return;
        }
        EnqueueDownloads(items, current);
    }

    /// Opens the folder picker, saves the choice for this type, and continues with it (null = app folder).
    public static void PickFolder(TransferType type, Action<string?> then)
    {
        var prefs = Prefs.Instance;
        var dlg = new Microsoft.Win32.OpenFolderDialog
        {
            Title = type == TransferType.RECORDING ? "Recordings folder" : "Download folder",
            InitialDirectory = (type == TransferType.RECORDING ? prefs.RecordingFolder : prefs.DownloadFolder) ?? Folders.AppDir(type)
        };
        var ok = App.Window != null ? dlg.ShowDialog(App.Window) : dlg.ShowDialog();
        if (ok == true)
        {
            var s = dlg.FolderName;
            if (type == TransferType.RECORDING) prefs.RecordingFolder = s; else prefs.DownloadFolder = s;
            then(s);
        }
        else if ((type == TransferType.RECORDING ? prefs.RecordingFolder : prefs.DownloadFolder) == null)
        {
            // Cancelled with nothing saved yet: use the app folder so the action still works.
            then(null);
        }
    }

    private static void EnqueueDownloads(IList<DownloadItem> items, string? folder)
    {
        foreach (var it in items)
        {
            var job = TransferJob.Create(TransferType.DOWNLOAD, it.Title, it.Subtitle, it.Url,
                Folders.SafeName(it.Title) + "." + (string.IsNullOrWhiteSpace(it.Ext) ? "mp4" : it.Ext), folder, 0, 0);
            TransferService.Enqueue(job);
        }
        Dialogs.Toast(items.Count == 1 ? $"Downloading {items[0].Title}." : $"Downloading {items.Count} items.");
    }

    /// Start and end clock times (to the minute), then folder and schedule.
    public static void Record(string channelName, string url)
    {
        var now = DateTime.Now;
        var startCal = now.AddMinutes(1);
        var endCal = startCal.AddHours(1);

        var panel = new StackPanel { Margin = new Thickness(4, 0, 4, 0), MinWidth = 360 };
        panel.Children.Add(new TextBlock { Text = "Start", Foreground = Ui.Brush("AccentBrush"), FontSize = 13 });
        var start = new TimeWheels(startCal.Hour, startCal.Minute);
        panel.Children.Add(start.View);
        panel.Children.Add(new TextBlock { Text = "End", Foreground = Ui.Brush("AccentBrush"), FontSize = 13, Margin = new Thickness(0, 8, 0, 0) });
        var end = new TimeWheels(endCal.Hour, endCal.Minute);
        panel.Children.Add(end.View);
        var summary = new TextBlock { Foreground = Ui.Brush("TextBrush"), FontSize = 14, Margin = new Thickness(0, 10, 0, 0), TextWrapping = TextWrapping.Wrap };
        panel.Children.Add(summary);
        var prefs = Prefs.Instance;
        panel.Children.Add(new TextBlock
        {
            Text = "Folder: " + Folders.Describe(Folders.Usable(prefs.RecordingFolder) ? prefs.RecordingFolder : null, TransferType.RECORDING),
            Foreground = Ui.Brush("TextMutedBrush"), FontSize = 13, Margin = new Thickness(0, 4, 0, 0), TextWrapping = TextWrapping.Wrap
        });

        (long, long) Times() => Resolve(start.Hour24, start.Minute, end.Hour24, end.Minute);
        void Summarize()
        {
            var (st, en) = Times();
            var s = DateTimeOffset.FromUnixTimeMilliseconds(st).LocalDateTime;
            var e = DateTimeOffset.FromUnixTimeMilliseconds(en).LocalDateTime;
            var sameDay = s.Date == e.Date;
            summary.Text = sameDay
                ? $"{s:ddd MMM d}   {s:h:mm tt} – {e:h:mm tt}"
                : $"{s:ddd MMM d} {s:h:mm tt} – {e:ddd MMM d} {e:h:mm tt}";
        }
        start.Changed += Summarize;
        end.Changed += Summarize;
        Summarize();

        var r = Dialogs.Custom($"Record {channelName}", panel, "Record", "Cancel", focus: start.HourBox);
        if (r != DialogResultKind.Positive) return;
        var (st2, en2) = Times();
        var current = Folders.Usable(prefs.RecordingFolder) ? prefs.RecordingFolder : null;
        if (current == null && prefs.RecordingFolder == null)
            PickFolder(TransferType.RECORDING, f => Schedule(channelName, url, st2, en2, f));
        else
            Schedule(channelName, url, st2, en2, current);
    }

    /// One hour box that runs through the whole day (12 AM … 11 AM, 12 PM … 11 PM) plus a minute box.
    private class TimeWheels
    {
        public readonly ComboBox HourBox = new() { Width = 110, Style = null };
        public readonly ComboBox MinuteBox = new() { Width = 80 };
        public readonly StackPanel View = new() { Orientation = Orientation.Horizontal, HorizontalAlignment = HorizontalAlignment.Center, Margin = new Thickness(0, 4, 0, 0) };
        public event Action? Changed;

        public TimeWheels(int hour24, int minute)
        {
            for (var h = 0; h < 24; h++)
            {
                var h12 = h % 12 == 0 ? 12 : h % 12;
                HourBox.Items.Add($"{h12} {(h < 12 ? "AM" : "PM")}");
            }
            for (var m = 0; m < 60; m++) MinuteBox.Items.Add(m.ToString("00", CultureInfo.InvariantCulture));
            HourBox.SelectedIndex = hour24;
            MinuteBox.SelectedIndex = minute;
            foreach (var cb in new[] { HourBox, MinuteBox })
            {
                cb.FontSize = 16;
                cb.MaxDropDownHeight = 260;
                cb.SelectionChanged += (_, _) => Changed?.Invoke();
            }
            View.Children.Add(HourBox);
            View.Children.Add(new TextBlock { Text = ":", Foreground = Ui.Brush("TextBrush"), FontSize = 22, Margin = new Thickness(8, 0, 8, 0), VerticalAlignment = VerticalAlignment.Center });
            View.Children.Add(MinuteBox);
        }

        public int Hour24 => Math.Max(0, HourBox.SelectedIndex);
        public int Minute => Math.Max(0, MinuteBox.SelectedIndex);
    }

    /// Turns start/end clock times into epoch millis: a start earlier than now means tomorrow
    /// (unless it is within the last minute, which means "now"), and an end at or before the
    /// start rolls over to the next day.
    private static (long, long) Resolve(int sh, int sm, int eh, int em)
    {
        var now = DateTimeOffset.Now;
        var start = new DateTimeOffset(now.Year, now.Month, now.Day, sh, sm, 0, now.Offset);
        var st = start.ToUnixTimeMilliseconds();
        var nowMs = now.ToUnixTimeMilliseconds();
        if (st < nowMs - 90_000) st += 24 * 3600 * 1000L;
        if (st < nowMs) st = nowMs;
        var stLocal = DateTimeOffset.FromUnixTimeMilliseconds(st).ToLocalTime();
        var end = new DateTimeOffset(stLocal.Year, stLocal.Month, stLocal.Day, eh, em, 0, stLocal.Offset);
        var en = end.ToUnixTimeMilliseconds();
        if (en <= st) en += 24 * 3600 * 1000L;
        return (st, en);
    }

    private static void Schedule(string channel, string url, long startAt, long endAt, string? folder)
    {
        var s = DateTimeOffset.FromUnixTimeMilliseconds(startAt).LocalDateTime;
        var e = DateTimeOffset.FromUnixTimeMilliseconds(endAt).LocalDateTime;
        var job = TransferJob.Create(TransferType.RECORDING, channel,
            $"{s:ddd MMM d} · {s:h:mm tt} – {e:h:mm tt}", url,
            Folders.SafeName($"{channel} {s:yyyy-MM-dd HH.mm}") + ".ts", folder, startAt, endAt);
        TransferService.Enqueue(job);
        var startsNow = startAt <= Format.NowMs + 60_000;
        Dialogs.Toast(startsNow ? $"Recording {channel}." : $"{channel} will be recorded at {s:h:mm tt}.");
    }
}
